lib/promscrape/discovery/kuma/api.go

```diff
@@ -0,0 +1,207 @@
+
+func getAPIConfig(sdc *SDConfig, baseDir string) (*apiConfig, error) {
+	v, err := configMap.Get(sdc, func() (interface{}, error) { return newAPIConfig(sdc, baseDir) })
+	if err != nil {
+		return nil, err
+	}
+	return v.(*apiConfig), nil
+}
+
+func newAPIConfig(sdc *SDConfig, baseDir string) (*apiConfig, error) {
+	ac, err := sdc.HTTPClientConfig.NewConfig(baseDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse auth config: %w", err)
+	}
+	parsedURL, err := url.Parse(sdc.Server)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse kuma_sd server URL: %w", err)
+	}
+	apiServer := fmt.Sprintf("%s://%s", parsedURL.Scheme, parsedURL.Host)
+
+	proxyAC, err := sdc.ProxyClientConfig.NewConfig(baseDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse proxy auth config: %w", err)
+	}
+	client, err := discoveryutils.NewClient(apiServer, ac, sdc.ProxyURL, proxyAC)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create HTTP client for %q: %w", apiServer, err)
+	}
+
+	apiPath := path.Join(
+		parsedURL.RequestURI(),
+		xdsApiVersion,
+		xdsRequestType+":"+xdsResourceType,
+	)
+
+	cfg := &apiConfig{
+		client:      client,
+		path:        apiPath,
+		fetchErrors: metrics.GetOrCreateCounter(fmt.Sprintf(`promscrape_discovery_kuma_errors_total{type="fetch",url=%q}`, sdc.Server)),
+		parseErrors: metrics.GetOrCreateCounter(fmt.Sprintf(`promscrape_discovery_kuma_errors_total{type="parse",url=%q}`, sdc.Server)),
+	}
+
+	// initialize targets synchronously and start updating them in background
+	cfg.startWatcher()
+
+	return cfg, nil
+}
+
+func (cfg *apiConfig) startWatcher() func() {
+	ctx, cancel := context.WithCancel(context.Background())
+	cfg.cancel = cancel
+
+	// blocking initial targets update
+	if err := cfg.updateTargets(ctx); err != nil {
+		logger.Errorf("there were errors when discovering kuma targets, so preserving the previous targets. error: %v", err)
+	}
+
+	// start updating targets with a long polling in background
+	cfg.wg.Add(1)
+	go func() {
+		ticker := time.NewTicker(*SDCheckInterval)
+		defer cfg.wg.Done()
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				// we are constantly waiting for targets updates in long polling requests
+				err := cfg.updateTargets(ctx)
+				if err != nil {
+					logger.Errorf("there were errors when discovering kuma targets, so preserving the previous targets. error: %v", err)
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return cancel
+}
+
+func (cfg *apiConfig) stopWatcher() {
+	cfg.cancel()
+	cfg.wg.Wait()
+}
+
+func (cfg *apiConfig) getTargets() ([]kumaTarget, error) {
+	cfg.mu.Lock()
+	defer cfg.mu.Unlock()
+	return cfg.targets, nil
+}
+
+func (cfg *apiConfig) updateTargets(ctx context.Context) error {
+	requestBody, err := json.Marshal(discoveryRequest{
+		VersionInfo:   cfg.latestVersion,
+		Node:          discoveryRequestNode{Id: discoveryNode},
```

Comment:
why do we use `victoria-metrics` as an ID here?

the promehteus implementation uses  the following code 
```
// Default to "prometheus" if hostname is unavailable.
	clientID, err := osutil.GetFQDN()
	if err != nil {
		level.Debug(logger).Log("msg", "error getting FQDN", "err", err)
		clientID = "prometheus"
	}
```

Reply:
I'm unsure whether the clientID is important in Kuma service discovery. It looks like some kind of synonym to User-Agent request header. In this case "victoria-metrics" string is OK to use as a clientID. BTW, I changed it to "vmagent" at 9fbd45a22fdba49fd496aa6d2760b1cd4e44a46f .

We need to change clientID to fully-qualified hostname of the server where vmagent runs as Prometheus does if Kuma server uses the clientID in some other tricky way.